File: samples/ios/iosExample/Merge/MergePage.swift
import SwiftUI
import AsyncStorageSQLiteKMP


struct MergePage: View {
    
    private let db = AsyncStorageSQLite(name: "merge.sqlite")
    private let defaultEntry: Entry
    
    @State var mergeValue = createMergingValue()
    @State var merged: String? = nil
    
    
    init() {
        defaultEntry = Entry(key: "default", value: defaultValue)
    }
    
    
    func randomizeMergeValue() {
        mergeValue = createMergingValue()
    }
    
    func reset() {
        merged = nil
        Task {
            try await db.write(entry: defaultEntry)
        }
    }
    
    func mergeValues() {
        Task {
            let updated = try await db.merge(entry: Entry(key: defaultEntry.key, value: mergeValue))
            merged = updated.value
        }
    }
    
    
    var body: some View {
        ScrollView {
            VStack {
                HStack(spacing: 20) {
                    Button("randomize", action: randomizeMergeValue)
                    Button("merge", role: .cancel, action: mergeValues)
                    Button("reset", role: .destructive, action: reset)
                }
                
                HStack {
                    Text(defaultEntry.value!.prettyJSON).font(.caption)
                    Divider()
                    Text(mergeValue.prettyJSON).font(.caption)
                }
                Spacer()
                if let merged = merged?.prettyJSON {
                    Divider()
                    Text("Merge result:").font(.headline)
                    Text(merged).font(.caption)
                }
                
            }
        }
    }
}


private let defaultValue = """
{
   "index":0,
   "isActive":false,
   "age":29,
   "name":"Johanna Sykes",
   "tags":[
      "tag-0",
      "tag-1",
      "tag-2"
   ],
   "properties":{
      "topIndex":0,
      "inner":false,
      "innerList":[
         1,
         2,
         3
      ],
      "innerProp":{
         "keep":"me",
         "name":"original"
      }
   }
}
"""

private func createMergingValue() -> String {
    let index = { Int.random(in: 0...5) }
    let boolean = { Bool.random() }
    let tags = {
        var t: [String] = []
        for _ in 0..<3 {
            t.append("tag-\(Int.random(in: 0...6))")
        }
        return t
    }
    let randomInt = { Int.random(in: 10...20)}
    
    return """
{
  "index": \(index()),
  "isActive": \(boolean()),
  "age": 21,
  "address": "\(Int.random(in: 4...9))th street",
  "tags": \(tags()),
  "properties": {
    "topIndex": \(index()),
    "inner": \(boolean()),
    "innerList": [
      \(randomInt()),
      \(randomInt()),
      \(randomInt())
    ],
    "innerProp": {
      "name": "overridden",
      "newProp": true
    }
  }
}
"""
}

extension String {
    var prettyJSON: String {
        let data = self.data(using: .utf8)!
        guard let object = try? JSONSerialization.jsonObject(with: data, options: []),
              let data = try? JSONSerialization.data(withJSONObject: object, options: [.prettyPrinted]),
              let prettyPrintedString = NSString(data: data, encoding: String.Encoding.utf8.rawValue) else { return "" }
        
        return String(prettyPrintedString)
    }
}

File: samples/android/src/main/java/com/asyncstorage/sqlite/android/example/ui/merge/MergePage.kt
package com.asyncstorage.sqlite.android.example.ui.merge

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Button
import androidx.compose.material.Divider
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.asyncstorage.sqlite.android.example.ui.rememberStorage
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import kotlinx.coroutines.launch
import org.asyncstorage.sqlitestorage.models.Entry
import kotlin.random.Random
import kotlin.random.nextInt

private val defaultKey = "default"

@Composable
fun MergePage(modifier: Modifier = Modifier) {
    val storage = rememberStorage(name = "merge-example.db")
    val scope = rememberCoroutineScope()
    var merged: String? by remember {
        mutableStateOf(null)
    }
    var mergeValue: String by remember {
        mutableStateOf(createMergingValue())
    }

    fun randomizeMergeValue() {
        mergeValue = createMergingValue()
    }


    fun reset() {
        merged = null
        scope.launch {
            storage.write(Entry(defaultKey, defaultValue))
        }
    }

    fun mergeValues() {
        scope.launch {
            val updated = storage.merge(Entry(key = defaultKey, value = mergeValue))
            merged = updated.value
        }
    }

    Column(
        modifier = modifier
            .verticalScroll(state = rememberScrollState())
            .padding(4.dp)
    ) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Button(onClick = ::randomizeMergeValue) {
                Text("randomize")
            }
            Button(onClick = ::mergeValues) {
                Text("merge")
            }
            Button(onClick = ::reset) {
                Text("Clear all", color = Color.Red)
            }
        }

        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Text(defaultValue.prettyJson(), fontSize = 12.sp)
            Text(mergeValue.prettyJson(), fontSize = 12.sp)
        }
        if (merged != null) {
            Divider(modifier = Modifier.padding(vertical = 12.dp))
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(text = merged!!.prettyJson())
            }
        }
    }
}


private fun createMergingValue(): String {
    val index = { Random.nextInt(0..5) }
    val boolean = { Random.nextBoolean() }
    val tags = {
        val t = mutableListOf<String>()
        for (i in 0..<3) {
            t.add("tag-${Random.nextInt(0..6)}")
        }
        t
    }
    val randomInt = { Random.nextInt(10..20) }

    return """
{
  "index": ${index()},
  "isActive": ${boolean()},
  "age": 21,
  "address": "${Random.nextInt(4..9)}th street",
  "tags": ${tags()},
  "properties": {
    "topIndex": ${index()},
    "inner": ${boolean()},
    "innerList": [
      ${randomInt()},
      ${randomInt()},
      ${randomInt()}
    ],
    "innerProp": {
      "name": "overridden",
      "newProp": true
    }
  }
}
"""
}

private val defaultValue = """
{
   "index":0,
   "isActive":false,
   "age":29,
   "name":"Johanna Sykes",
   "tags":[
      "tag-0",
      "tag-1",
      "tag-2"
   ],
   "properties":{
      "topIndex":0,
      "inner":false,
      "innerList":[
         1,
         2,
         3
      ],
      "innerProp":{
         "keep":"me",
         "name":"original"
      }
   }
}
"""

private fun String.prettyJson(): String {
    val obj = JsonParser.parseString(this).asJsonObject
    val gson = GsonBuilder().setPrettyPrinting().create()
    return gson.toJson(obj)
}
